accepted/2021/System.Diagnostics/Metrics-Design.md

```diff
@@ -0,0 +1,502 @@
+                            string name,
+                            Func<T> observeValue,
+                            string? description = null,
+                            string? unit = null) where T : unmanaged { throw null; }
+
+        public ObservableUpDownCounter<T> CreateObservableUpDownCounter<T>(
+                            string name,
+                            Func<Measurement<T>> observeValues,
+                            string? description = null,
+                            string? unit = null) where T : unmanaged { throw null; }
+
+        public ObservableUpDownCounter<T> CreateObservableUpDownCounter<T>(
+                            string name,
+                            Func<IEnumerable<Measurement<T>>> observeValues,
+                            string? description = null,
+                            string? unit = null) where T : unmanaged { throw null; }
+
+        public void Dispose() { throw null; }
+    }
+}
+```
+
+### Instrument Base Class
+
+```csharp
+namespace System.Diagnostics.Metrics
+{
+    /// <summary>
+    /// Is the base class which contains all common properties between different types of instruments.
+    /// It contains the protected constructor and the Publish method allows activating the instrument
+    /// to start recording measurements.
+    /// </summary>
+
+    public abstract class Instrument
+    {
+        /// <summary>
+        /// Protected constructor to initialize the common instrument properties.
+        /// </summary>
+        protected Instrument(Meter meter, string name, string? description, string? unit) { throw null; }
+
+        /// <summary>
+        /// Publish is to allow activating the instrument to start recording measurements and to allow
+        /// listeners to start listening to such measurements.
+        /// </summary>
+        protected void Publish() { throw null; }
+
+        /// <summary>
+        /// Getters to retrieve the properties that the instrument is created with.
+        /// </summary>
+        public Meter Meter { get; }
+        public string Name { get; }
+        public string? Description { get; }
+        public string? Unit { get; }
+
+        /// <summary>
+        /// A property tells if a listener is listening to this instrument measurement recording.
+        /// </summary>
+        public bool Enabled => throw null;
+
+        /// <summary>
+        /// A property tells if the instrument is a regular instrument or an observable instrument.
+        /// </summary>
+        public virtual bool IsObservable => throw null;
+    }
+}
+```
+
+### Non-observable Instrument Base Class
+
+```csharp
+namespace System.Diagnostics.Metrics
+{
+    /// <summary>
+    /// Instrument<T> is the base class from which all non-observable instruments will inherit from.
+    /// Mainly It'll support the CLS compliant numerical types
+    /// </summary>
+    public abstract class Instrument<T> : Instrument where T : unmanaged
```

Comment:
Where does the validation for T occurs? Will api validate that it'll be one of the numerical types? Or is that left to the SDK?

Reply:
the validation will be done in the API level and not the SDK.

Reply:
How do you intend to report errors?

Reply:
I think Exception throwing for non-supported types would be reasonable here. I don't expect someone will use some non-supported type without testing it :-)